File: python/confident_agent/schemas.py
from enum import Enum
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict


class ResponseMode(str, Enum):
    HTTP_RESPONSE = "HTTP_RESPONSE"
    SSE_STREAMING = "SSE_STREAMING"
    HTTP_STREAMING = "HTTP_STREAMING"


class RelayRequest(BaseModel):
    id: str
    method: str
    url: str
    headers: Dict[str, str] = {}
    body: Optional[Any] = None
    timeout: int = 60
    response_mode: ResponseMode = ResponseMode.HTTP_RESPONSE


class RelayResponseType(str, Enum):
    RESPONSE = "response"
    ERROR = "error"
    CHUNK = "chunk"
    STREAM_END = "stream_end"


class RelayResponse(BaseModel):
    id: str
    type: RelayResponseType
    statusCode: int
    headers: Dict[str, str] = {}
    body: str


class ToolCall(BaseModel):
    model_config = ConfigDict(extra="allow")

    name: str
    description: Optional[str] = None
    input_parameters: Optional[Dict[str, Any]] = None
    output: Optional[Any] = None


class Turn(BaseModel):
    model_config = ConfigDict(extra="allow")

    role: str
    content: str


class Request(BaseModel):
    model_config = ConfigDict(extra="allow")

    input: str = ""
    context: List[str] = []
    retrieval_context: List[str] = []
    expected_output: Optional[str] = None
    turns: List[Turn] = []
    scenario: Optional[str] = None
    state: Any = None
    prompts: Optional[Dict[str, Any]] = None
    hyperparameters: Optional[Dict[str, Any]] = None
    test_case_id: Optional[str] = None
    turn_id: Optional[str] = None


class Response(BaseModel):
    output: str
    retrieval_context: Optional[List[str]] = None
    tools_called: Optional[List[ToolCall]] = None
    state: Any = None

File: rust/src/schemas.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum ResponseMode {
    #[default]
    HttpResponse,
    SseStreaming,
    HttpStreaming,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RelayRequest {
    pub id: String,
    pub method: String,
    pub url: String,
    #[serde(default)]
    pub headers: HashMap<String, String>,
    #[serde(default)]
    pub body: Option<Value>,
    #[serde(default = "default_timeout")]
    pub timeout: u64,
    #[serde(default)]
    pub response_mode: ResponseMode,
}

fn default_timeout() -> u64 {
    60
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum RelayResponseType {
    Response,
    Error,
    Chunk,
    StreamEnd,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RelayResponse {
    pub id: String,
    pub r#type: RelayResponseType,
    #[serde(rename = "statusCode")]
    pub status_code: u16,
    #[serde(default)]
    pub headers: HashMap<String, String>,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolCall {
    pub name: String,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub input_parameters: Option<Value>,
    #[serde(default)]
    pub output: Option<Value>,
    #[serde(flatten)]
    pub extra: HashMap<String, Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Turn {
    pub role: String,
    pub content: String,
    #[serde(flatten)]
    pub extra: HashMap<String, Value>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct Request {
    #[serde(default)]
    pub input: String,
    #[serde(default)]
    pub context: Vec<String>,
    #[serde(default)]
    pub retrieval_context: Vec<String>,
    #[serde(default)]
    pub expected_output: Option<String>,
    #[serde(default)]
    pub turns: Vec<Turn>,
    #[serde(default)]
    pub scenario: Option<String>,
    #[serde(default)]
    pub state: Option<Value>,
    #[serde(default)]
    pub prompts: Option<Value>,
    #[serde(default)]
    pub hyperparameters: Option<Value>,
    #[serde(default)]
    pub test_case_id: Option<String>,
    #[serde(default)]
    pub turn_id: Option<String>,
    #[serde(flatten)]
    pub extra: HashMap<String, Value>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct Response {
    pub output: String,
    #[serde(default)]
    pub retrieval_context: Option<Vec<String>>,
    #[serde(default)]
    pub tools_called: Option<Vec<ToolCall>>,
    #[serde(default)]
    pub state: Option<Value>,
}

impl From<String> for Response {
    fn from(output: String) -> Self {
        Response {
            output,
            ..Response::default()
        }
    }
}

impl From<&str> for Response {
    fn from(output: &str) -> Self {
        Response::from(output.to_string())
    }
}
